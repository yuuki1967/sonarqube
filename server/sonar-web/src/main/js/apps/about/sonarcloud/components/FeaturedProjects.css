/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.sc-featured-projects {
  margin-top: 40px;
  position: relative;
}

.sc-featured-projects-container {
  margin: 0 40px;
  overflow: hidden;
  padding: 10px 8px 20px;
  max-width: 985px;
}

.sc-featured-projects-inner {
  display: flex;
  order: 2;
  position: relative;
  left: calc(100% / -3);
  transform: translateX(calc(100% / 3));
}

.sc-featured-projects-inner.reversing {
  transform: translateX(calc(100% / -3));
}

.sc-featured-projects-inner.ready {
  transform: translateX(0);
  transition: transform 0.5s cubic-bezier(0.23, 1, 0.32, 1);
}

.sc-project-card-container {
  flex: 1 0 calc(100% / 3);
}

.sc-project-card {
  width: 255px;
  display: inline-block;
  margin: 0 10px;
  padding: 25px;
  border: 1px solid var(--sonarcloudBlack300) !important;
  border-radius: 5px;
  box-shadow: 0 1px 1px rgba(7, 7, 6, 0.1);
  transition: all 0.1s ease-in;
  color: inherit;
}

.sc-project-card:hover,
.sc-project-card:focus {
  box-shadow: 0 10px 30px rgba(7, 7, 6, 0.2);
  transform: translateY(-4px);
  color: inherit;
}

.sc-project-card-header {
  padding: 0 10px 16px;
}

.sc-project-card-limited {
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.sc-project-card-measures li {
  display: flex;
  align-items: baseline;
  justify-content: space-between;
  padding: 6px 0 4px;
  border-bottom: 1px solid var(--sonarcloudBlack250);
}

.sc-project-card-measures li:last-of-type {
  border: none;
}

.sc-project-card-measures li div {
  display: inline-flex;
}

.sc-project-card .sc-mention {
  color: var(--sonarcloudBlack500);
}

.sc-project-button {
  display: block;
  position: absolute;
  top: 50%;
  margin: 0;
  padding: 4px;
  border: none;
  background: transparent;
  color: var(--sonarcloudBlack300);
  text-decoration: none;
  cursor: pointer;
  outline: none;
  transition: all 0.2s ease;
}

.sc-project-button img {
  opacity: 0.5;
  width: 19px;
}

.sc-project-button:hover img {
  opacity: 1;
}

.sc-project-button:last-child {
  right: 0;
}

.sc-project-button:hover,
.sc-project-button:active,
.sc-project-button:focus {
  color: var(--sonarcloudBlack500);
}
