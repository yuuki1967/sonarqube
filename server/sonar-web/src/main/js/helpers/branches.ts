/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { sortBy } from 'lodash';

export function isBranch(branchLike?: T.BranchLike): branchLike is T.Branch {
  return branchLike !== undefined && (branchLike as T.Branch).isMain !== undefined;
}

export function isShortLivingBranch(branchLike?: T.BranchLike): branchLike is T.ShortLivingBranch {
  return (
    isBranch(branchLike) &&
    !branchLike.isMain &&
    (branchLike as T.ShortLivingBranch).type === 'SHORT'
  );
}

export function isLongLivingBranch(branchLike?: T.BranchLike): branchLike is T.LongLivingBranch {
  return (
    isBranch(branchLike) && !branchLike.isMain && (branchLike as T.LongLivingBranch).type === 'LONG'
  );
}

export function isMainBranch(branchLike?: T.BranchLike): branchLike is T.MainBranch {
  return isBranch(branchLike) && branchLike.isMain;
}

export function isPullRequest(branchLike?: T.BranchLike): branchLike is T.PullRequest {
  return branchLike !== undefined && (branchLike as T.PullRequest).key !== undefined;
}

export function getPullRequestDisplayName(pullRequest: T.PullRequest) {
  return `${pullRequest.key} â€“ ${pullRequest.title}`;
}

export function getBranchLikeDisplayName(branchLike: T.BranchLike) {
  return isPullRequest(branchLike) ? getPullRequestDisplayName(branchLike) : branchLike.name;
}

export function getBranchLikeKey(branchLike: T.BranchLike) {
  return isPullRequest(branchLike) ? `pull-request-${branchLike.key}` : `branch-${branchLike.name}`;
}

export function getBranchQualityGateColor(status: string) {
  let indicatorColor = 'gray';
  if (status === 'ERROR') {
    indicatorColor = 'red';
  } else if (status === 'WARN') {
    indicatorColor = 'orange';
  } else if (status === 'OK') {
    indicatorColor = 'green';
  }
  return indicatorColor;
}

export function isSameBranchLike(a: T.BranchLike | undefined, b: T.BranchLike | undefined) {
  // main branches are always equal
  if (isMainBranch(a) && isMainBranch(b)) {
    return true;
  }

  // short- and long-living branches are compared by type and name
  if (
    (isLongLivingBranch(a) && isLongLivingBranch(b)) ||
    (isShortLivingBranch(a) && isShortLivingBranch(b))
  ) {
    return a.type === b.type && a.name === b.name;
  }

  // pull requests are compared by id
  if (isPullRequest(a) && isPullRequest(b)) {
    return a.key === b.key;
  }

  // finally if both parameters are `undefined`, consider them equal
  return a === b;
}

export function sortBranchesAsTree(branchLikes: T.BranchLike[]) {
  const result: T.BranchLike[] = [];

  const mainBranch = branchLikes.find(isMainBranch);
  const longLivingBranches = branchLikes.filter(isLongLivingBranch);
  const shortLivingBranches = branchLikes.filter(isShortLivingBranch);
  const pullRequests = branchLikes.filter(isPullRequest);

  // main branch is always first
  if (mainBranch) {
    result.push(
      mainBranch,
      ...getPullRequests(mainBranch.name),
      ...getNestedShortLivingBranches(mainBranch.name)
    );
  }

  // then all long-living branches
  sortBy(longLivingBranches, 'name').forEach(longLivingBranch => {
    result.push(
      longLivingBranch,
      ...getPullRequests(longLivingBranch.name),
      ...getNestedShortLivingBranches(longLivingBranch.name)
    );
  });

  // finally all orhpan pull requests and branches
  result.push(
    ...sortBy(pullRequests.filter(pr => pr.isOrphan), pullRequest => pullRequest.key),
    ...sortBy(shortLivingBranches.filter(branch => branch.isOrphan), branch => branch.name)
  );

  return result;

  /** Get all short-living branches (possibly nested) which should be merged to a given branch */
  function getNestedShortLivingBranches(mergeBranch: string) {
    const found: T.ShortLivingBranch[] = shortLivingBranches.filter(
      branch => branch.mergeBranch === mergeBranch
    );

    let i = 0;
    while (i < found.length) {
      const current = found[i];
      found.push(...shortLivingBranches.filter(branch => branch.mergeBranch === current.name));
      i++;
    }

    return sortBy(found, branch => branch.name);
  }

  function getPullRequests(base: string) {
    return sortBy(pullRequests.filter(pr => pr.base === base), pullRequest => pullRequest.key);
  }
}

export function getBranchLikeQuery(branchLike?: T.BranchLike): T.BranchParameters {
  if (isShortLivingBranch(branchLike) || isLongLivingBranch(branchLike)) {
    return { branch: branchLike.name };
  } else if (isPullRequest(branchLike)) {
    return { pullRequest: branchLike.key };
  } else {
    return {};
  }
}

// Create branch object from branch name or pull request key
export function fillBranchLike(
  branch?: string,
  pullRequest?: string
): T.ShortLivingBranch | T.PullRequest | undefined {
  if (branch) {
    return {
      isMain: false,
      mergeBranch: '',
      name: branch,
      type: 'SHORT'
    } as T.ShortLivingBranch;
  } else if (pullRequest) {
    return { base: '', branch: '', key: pullRequest, title: '' } as T.PullRequest;
  }
  return undefined;
}
