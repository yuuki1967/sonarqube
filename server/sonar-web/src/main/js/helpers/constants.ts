/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as theme from '../app/theme';

export const SEVERITIES = ['BLOCKER', 'CRITICAL', 'MAJOR', 'MINOR', 'INFO'];
export const STATUSES = ['OPEN', 'REOPENED', 'CONFIRMED', 'RESOLVED', 'CLOSED'];
export const ISSUE_TYPES: T.IssueType[] = [
  'BUG',
  'VULNERABILITY',
  'CODE_SMELL',
  'SECURITY_HOTSPOT'
];
export const RULE_TYPES: T.RuleType[] = [
  'BUG',
  'VULNERABILITY',
  'CODE_SMELL',
  'SECURITY_HOTSPOT',
  'UNKNOWN'
];
export const RULE_STATUSES = ['READY', 'BETA', 'DEPRECATED'];

export const CHART_COLORS_RANGE_PERCENT = [
  theme.green,
  theme.lightGreen,
  theme.yellow,
  theme.orange,
  theme.red
];
export const CHART_REVERSED_COLORS_RANGE_PERCENT = [
  theme.red,
  theme.orange,
  theme.yellow,
  theme.lightGreen,
  theme.green
];

export const RATING_COLORS = [theme.green, theme.lightGreen, theme.yellow, theme.orange, '#e00'];
