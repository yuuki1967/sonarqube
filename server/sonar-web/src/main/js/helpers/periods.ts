/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { translate, translateWithParameters } from './l10n';
import { parseDate } from './dates';

function getPeriod<T extends T.Period | T.PeriodMeasure>(periods: T[] | undefined, index: number) {
  if (!Array.isArray(periods)) {
    return undefined;
  }
  return periods.find(period => period.index === index);
}

export function getLeakPeriod<T extends T.Period | T.PeriodMeasure>(periods: T[] | undefined) {
  return getPeriod(periods, 1);
}

export function getPeriodLabel(
  period: T.Period | undefined,
  dateFormatter: (date: string) => string
) {
  if (!period) {
    return undefined;
  }

  let parameter = period.modeParam || period.parameter;
  if (period.mode === 'previous_version' && !parameter) {
    return translate('overview.period.previous_version_only_date');
  }

  if (period.mode === 'date' && parameter) {
    parameter = dateFormatter(parameter);
  } else if (period.mode === 'manual_baseline') {
    if (!parameter) {
      parameter = dateFormatter(period.date);
    } else {
      return translateWithParameters('overview.period.previous_version', parameter);
    }
  }

  return translateWithParameters(`overview.period.${period.mode}`, parameter || '');
}

export function getPeriodDate(period?: { date?: string }): Date | undefined {
  return period && period.date ? parseDate(period.date) : undefined;
}
