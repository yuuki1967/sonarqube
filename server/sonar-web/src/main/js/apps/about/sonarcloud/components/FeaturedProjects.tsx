/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import CountUp from 'react-countup';
import { throttle } from 'lodash';
import { FeaturedProject } from '../utils';
import CoverageRating from '../../../../components/ui/CoverageRating';
import DuplicationsRating from '../../../../components/ui/DuplicationsRating';
import OrganizationAvatar from '../../../../components/common/OrganizationAvatar';
import ProjectCardLanguagesContainer from '../../../projects/components/ProjectCardLanguagesContainer';
import Rating from '../../../../components/ui/Rating';
import { formatMeasure } from '../../../../helpers/measures';
import { getMetricName } from '../../../overview/utils';
import { getProjectUrl, getBaseUrl, getPathUrlAsString } from '../../../../helpers/urls';
import './FeaturedProjects.css';

interface Props {
  projects: FeaturedProject[];
}

interface State {
  reversing: boolean;
  slides: Array<{
    order: number;
    project: FeaturedProject;
  }>;
  sliding: boolean;
  translate: number;
  viewable: boolean;
}

export default class FeaturedProjects extends React.PureComponent<Props, State> {
  container?: HTMLElement | null;
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      reversing: false,
      slides: this.orderProjectsFromProps(),
      sliding: false,
      translate: 0,
      viewable: false
    };
    this.handleScroll = throttle(this.handleScroll, 10);
  }

  componentDidMount() {
    this.mounted = true;
    document.addEventListener('scroll', this.handleScroll, true);
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.projects !== this.props.projects) {
      this.setState({ slides: this.orderProjectsFromProps() });
    }
  }

  componentWillUnmount() {
    this.mounted = false;
    document.removeEventListener('scroll', this.handleScroll, true);
  }

  handleScroll = () => {
    if (this.container) {
      const rect = this.container.getBoundingClientRect();
      const windowHeight =
        window.innerHeight ||
        (document.documentElement ? document.documentElement.clientHeight : 0);
      if (rect.top <= windowHeight && rect.top + rect.height >= 0) {
        this.setState({ viewable: true });
      }
    }
  };

  orderProjectsFromProps = () => {
    const { projects } = this.props;
    if (projects.length === 0) {
      return [];
    }

    // Last element should be put at the begining for proper carousel animation
    return [projects.pop(), ...projects].map((project: FeaturedProject, id) => {
      return {
        order: id,
        project
      };
    });
  };

  handlePrevClick = () => {
    this.setState(({ slides }) => ({
      reversing: true,
      sliding: true,
      slides: slides.map(slide => {
        slide.order = slide.order === slides.length - 1 ? 0 : slide.order + 1;
        return slide;
      })
    }));
    setTimeout(() => {
      if (this.mounted) {
        this.setState({ sliding: false });
      }
    }, 50);
  };

  handleNextClick = () => {
    this.setState(({ slides }) => ({
      reversing: false,
      sliding: true,
      slides: slides.map(slide => {
        slide.order = slide.order === 0 ? slides.length - 1 : slide.order - 1;
        return slide;
      })
    }));
    setTimeout(() => {
      this.setState({ sliding: false });
    }, 50);
  };

  render() {
    const { reversing, sliding, viewable } = this.state;
    return (
      <div
        className="sc-featured-projects sc-big-spacer-bottom"
        ref={node => (this.container = node)}>
        <button className="js-prev sc-project-button" onClick={this.handlePrevClick} type="button">
          <img alt="" src={`${getBaseUrl()}/images/sonarcloud/chevron-left.svg`} />
        </button>

        <div className="sc-featured-projects-container">
          <div
            className={classNames('sc-featured-projects-inner', {
              reversing,
              ready: !sliding
            })}>
            {this.state.slides.map(slide => (
              <ProjectCard
                key={slide.project.key}
                order={slide.order}
                project={slide.project}
                viewable={viewable}
              />
            ))}
          </div>
        </div>

        <button className="js-next sc-project-button" onClick={this.handleNextClick} type="button">
          <img alt="" src={`${getBaseUrl()}/images/sonarcloud/chevron-right.svg`} />
        </button>
      </div>
    );
  }
}

interface ProjectCardProps {
  order: number;
  project: FeaturedProject;
  viewable: boolean;
}

export function ProjectCard({ project, order, viewable }: ProjectCardProps) {
  return (
    <div className="sc-project-card-container" style={{ order }}>
      <a className="sc-project-card" href={getPathUrlAsString(getProjectUrl(project.key))}>
        <div className="sc-project-card-header">
          <OrganizationAvatar
            className="no-border spacer-bottom"
            organization={{
              name: project.organizationName,
              avatar: project.avatarUrl || undefined
            }}
          />
          <p className="sc-project-card-limited" title={project.organizationName}>
            {project.organizationName}
          </p>
          <h5 className="sc-project-card-limited big-spacer-bottom" title={project.name}>
            {project.name}
          </h5>
        </div>
        <ul className="sc-project-card-measures">
          <ProjectIssues
            metric={project.bugs}
            metricKey="bugs"
            ratingMetric={project.reliabilityRating}
            viewable={viewable}
          />
          <ProjectIssues
            metric={project.vulnerabilities}
            metricKey="vulnerabilities"
            ratingMetric={project.securityRating}
            viewable={viewable}
          />
          <ProjectIssues
            metric={project.codeSmells}
            metricKey="code_smells"
            ratingMetric={project.maintainabilityRating}
            viewable={viewable}
          />
          <li>
            <span>{getMetricName('coverage')}</span>
            <div>
              {viewable && (
                <CountUp
                  decimal="."
                  decimals={1}
                  delay={0}
                  duration={4}
                  end={project.coverage}
                  suffix="%">
                  {(data: { countUpRef?: React.RefObject<HTMLHeadingElement> }) => (
                    <h6 className="display-inline-block big-spacer-right" ref={data.countUpRef}>
                      0
                    </h6>
                  )}
                </CountUp>
              )}
              <CoverageRating value={project.coverage} />
            </div>
          </li>
          <li>
            <span>{getMetricName('duplications')}</span>
            <div>
              {viewable && (
                <CountUp
                  decimal="."
                  decimals={1}
                  delay={0}
                  duration={4}
                  end={project.duplications}
                  suffix="%">
                  {(data: { countUpRef?: React.RefObject<HTMLHeadingElement> }) => (
                    <h6 className="display-inline-block big-spacer-right" ref={data.countUpRef}>
                      0
                    </h6>
                  )}
                </CountUp>
              )}
              <DuplicationsRating value={project.duplications} />
            </div>
          </li>
        </ul>
        <div className="sc-mention text-left big-spacer-top">
          {formatMeasure(project.ncloc, 'SHORT_INT')} lines of code /{' '}
          <ProjectCardLanguagesContainer
            className="display-inline-block"
            distribution={project.languages.join(';')}
          />
        </div>
      </a>
    </div>
  );
}

interface ProjectIssues {
  metricKey: string;
  metric: number;
  ratingMetric: number;
  viewable: boolean;
}

export function ProjectIssues({ metric, metricKey, ratingMetric, viewable }: ProjectIssues) {
  const formattedValue = formatMeasure(metric, 'SHORT_INT');
  const value = parseFloat(formattedValue);
  const suffix = formattedValue.replace(value.toString(), '');
  return (
    <li>
      <span>{getMetricName(metricKey)}</span>
      <div>
        {viewable && (
          <CountUp delay={0} duration={4} end={value} suffix={suffix}>
            {(data: { countUpRef?: React.RefObject<HTMLHeadingElement> }) => (
              <h6 className="display-inline-block big-spacer-right" ref={data.countUpRef}>
                0
              </h6>
            )}
          </CountUp>
        )}
        <Rating value={ratingMetric} />
      </div>
    </li>
  );
}
