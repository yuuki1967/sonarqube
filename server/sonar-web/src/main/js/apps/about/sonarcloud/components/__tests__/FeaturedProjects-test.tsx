/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import FeaturedProjects, { ProjectCard, ProjectIssues } from '../FeaturedProjects';
import { click } from '../../../../../helpers/testUtils';

const PROJECTS = [
  {
    key: 'sonarsource-jfrog.simple-js-php-project',
    avatarUrl: null,
    organizationKey: 'sonarsource-jfrog',
    organizationName: 'SonarSource & JFrog',
    name: 'Simple JS & PHP project',
    bugs: 0,
    codeSmells: 7,
    coverage: 9.7,
    duplications: 56.2,
    gateStatus: 'OK',
    languages: ['js', 'php'],
    maintainabilityRating: 1,
    ncloc: 324,
    reliabilityRating: 1,
    securityRating: 1,
    vulnerabilities: 0
  },
  {
    key: 'example-js',
    avatarUrl: null,
    organizationKey: 'autoscan',
    organizationName: 'AutoScan',
    name: 'example-js',
    bugs: 13,
    codeSmells: 5,
    coverage: 0,
    duplications: 0,
    gateStatus: 'OK',
    languages: ['go', 'js', 'php', 'py'],
    maintainabilityRating: 1,
    ncloc: 80,
    reliabilityRating: 1,
    securityRating: 1,
    vulnerabilities: 0
  },
  {
    key: 'example-js-2',
    avatarUrl: null,
    organizationKey: 'autoscan',
    organizationName: 'AutoScan',
    name: 'example-js',
    bugs: 13,
    codeSmells: 5,
    coverage: 0,
    duplications: 0,
    gateStatus: 'OK',
    languages: ['go', 'js', 'php', 'py'],
    maintainabilityRating: 1,
    ncloc: 80,
    reliabilityRating: 1,
    securityRating: 1,
    vulnerabilities: 0
  },
  {
    key: 'example-js-3',
    avatarUrl: null,
    organizationKey: 'autoscan',
    organizationName: 'AutoScan',
    name: 'example-js',
    bugs: 13,
    codeSmells: 5,
    coverage: 0,
    duplications: 0,
    gateStatus: 'OK',
    languages: ['go', 'js', 'php', 'py'],
    maintainabilityRating: 1,
    ncloc: 80,
    reliabilityRating: 1,
    securityRating: 1,
    vulnerabilities: 0
  },
  {
    key: 'example-js-4',
    avatarUrl: null,
    organizationKey: 'autoscan',
    organizationName: 'AutoScan',
    name: 'example-js',
    bugs: 13,
    codeSmells: 5,
    coverage: 0,
    duplications: 0,
    gateStatus: 'OK',
    languages: ['go', 'js', 'php', 'py'],
    maintainabilityRating: 1,
    ncloc: 80,
    reliabilityRating: 1,
    securityRating: 1,
    vulnerabilities: 0
  }
];

it('should render ProjectIssues correctly', () => {
  expect(
    shallow(<ProjectIssues metric={5} metricKey="foo" ratingMetric={20} viewable={true} />)
  ).toMatchSnapshot();
  expect(
    shallow(<ProjectIssues metric={15000} metricKey="foo" ratingMetric={20} viewable={true} />)
  ).toMatchSnapshot();
});

it('should render ProjectCard correctly', () => {
  expect(
    shallow(<ProjectCard order={1} project={PROJECTS[0]} viewable={true} />)
  ).toMatchSnapshot();
});

it('should render correctly', () => {
  const wrapper = shallow(<FeaturedProjects projects={PROJECTS} />);
  expect(wrapper).toMatchSnapshot();
});

it('should cycle through projects', () => {
  const wrapper = shallow<FeaturedProjects>(<FeaturedProjects projects={PROJECTS} />);
  expect(wrapper.state().slides.map((slide: any) => slide.order)).toEqual([0, 1, 2, 3]);

  click(wrapper.find('.js-next'));
  expect(wrapper.state().slides.map((slide: any) => slide.order)).toEqual([3, 0, 1, 2]);

  click(wrapper.find('.js-next'));
  click(wrapper.find('.js-next'));
  expect(wrapper.state().slides.map((slide: any) => slide.order)).toEqual([1, 2, 3, 0]);

  click(wrapper.find('.js-prev'));
  click(wrapper.find('.js-prev'));
  expect(wrapper.state().slides.map((slide: any) => slide.order)).toEqual([3, 0, 1, 2]);
});
