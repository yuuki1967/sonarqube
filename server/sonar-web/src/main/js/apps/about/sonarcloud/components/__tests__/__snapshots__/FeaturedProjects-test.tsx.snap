// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ProjectCard correctly 1`] = `
<div
  className="sc-project-card-container"
  style={
    Object {
      "order": 1,
    }
  }
>
  <a
    className="sc-project-card"
    href="/dashboard?id=sonarsource-jfrog.simple-js-php-project"
  >
    <div
      className="sc-project-card-header"
    >
      <OrganizationAvatar
        className="no-border spacer-bottom"
        organization={
          Object {
            "avatar": undefined,
            "name": "SonarSource & JFrog",
          }
        }
      />
      <p
        className="sc-project-card-limited"
        title="SonarSource & JFrog"
      >
        SonarSource & JFrog
      </p>
      <h5
        className="sc-project-card-limited big-spacer-bottom"
        title="Simple JS & PHP project"
      >
        Simple JS & PHP project
      </h5>
    </div>
    <ul
      className="sc-project-card-measures"
    >
      <ProjectIssues
        metric={0}
        metricKey="bugs"
        ratingMetric={1}
        viewable={true}
      />
      <ProjectIssues
        metric={0}
        metricKey="vulnerabilities"
        ratingMetric={1}
        viewable={true}
      />
      <ProjectIssues
        metric={7}
        metricKey="code_smells"
        ratingMetric={1}
        viewable={true}
      />
      <li>
        <span>
          overview.metric.coverage
        </span>
        <div>
          <CountUp
            decimal="."
            decimals={1}
            delay={0}
            duration={4}
            end={9.7}
            suffix="%"
          >
            <Component />
          </CountUp>
          <CoverageRating
            value={9.7}
          />
        </div>
      </li>
      <li>
        <span>
          overview.metric.duplications
        </span>
        <div>
          <CountUp
            decimal="."
            decimals={1}
            delay={0}
            duration={4}
            end={56.2}
            suffix="%"
          >
            <Component />
          </CountUp>
          <DuplicationsRating
            value={56.2}
          />
        </div>
      </li>
    </ul>
    <div
      className="sc-mention text-left big-spacer-top"
    >
      324
       lines of code /
       
      <Connect(ProjectCardLanguages)
        className="display-inline-block"
        distribution="js;php"
      />
    </div>
  </a>
</div>
`;

exports[`should render ProjectIssues correctly 1`] = `
<li>
  <span>
    overview.metric.foo
  </span>
  <div>
    <CountUp
      delay={0}
      duration={4}
      end={5}
      suffix=""
    >
      <Component />
    </CountUp>
    <Rating
      value={20}
    />
  </div>
</li>
`;

exports[`should render ProjectIssues correctly 2`] = `
<li>
  <span>
    overview.metric.foo
  </span>
  <div>
    <CountUp
      delay={0}
      duration={4}
      end={15}
      suffix="short_number_suffix.k"
    >
      <Component />
    </CountUp>
    <Rating
      value={20}
    />
  </div>
</li>
`;

exports[`should render correctly 1`] = `
<div
  className="sc-featured-projects sc-big-spacer-bottom"
>
  <button
    className="js-prev sc-project-button"
    onClick={[Function]}
    type="button"
  >
    <img
      alt=""
      src="/images/sonarcloud/chevron-left.svg"
    />
  </button>
  <div
    className="sc-featured-projects-container"
  >
    <div
      className="sc-featured-projects-inner ready"
    >
      <ProjectCard
        key="example-js-4"
        order={0}
        project={
          Object {
            "avatarUrl": null,
            "bugs": 13,
            "codeSmells": 5,
            "coverage": 0,
            "duplications": 0,
            "gateStatus": "OK",
            "key": "example-js-4",
            "languages": Array [
              "go",
              "js",
              "php",
              "py",
            ],
            "maintainabilityRating": 1,
            "name": "example-js",
            "ncloc": 80,
            "organizationKey": "autoscan",
            "organizationName": "AutoScan",
            "reliabilityRating": 1,
            "securityRating": 1,
            "vulnerabilities": 0,
          }
        }
        viewable={false}
      />
      <ProjectCard
        key="sonarsource-jfrog.simple-js-php-project"
        order={1}
        project={
          Object {
            "avatarUrl": null,
            "bugs": 0,
            "codeSmells": 7,
            "coverage": 9.7,
            "duplications": 56.2,
            "gateStatus": "OK",
            "key": "sonarsource-jfrog.simple-js-php-project",
            "languages": Array [
              "js",
              "php",
            ],
            "maintainabilityRating": 1,
            "name": "Simple JS & PHP project",
            "ncloc": 324,
            "organizationKey": "sonarsource-jfrog",
            "organizationName": "SonarSource & JFrog",
            "reliabilityRating": 1,
            "securityRating": 1,
            "vulnerabilities": 0,
          }
        }
        viewable={false}
      />
      <ProjectCard
        key="example-js"
        order={2}
        project={
          Object {
            "avatarUrl": null,
            "bugs": 13,
            "codeSmells": 5,
            "coverage": 0,
            "duplications": 0,
            "gateStatus": "OK",
            "key": "example-js",
            "languages": Array [
              "go",
              "js",
              "php",
              "py",
            ],
            "maintainabilityRating": 1,
            "name": "example-js",
            "ncloc": 80,
            "organizationKey": "autoscan",
            "organizationName": "AutoScan",
            "reliabilityRating": 1,
            "securityRating": 1,
            "vulnerabilities": 0,
          }
        }
        viewable={false}
      />
      <ProjectCard
        key="example-js-2"
        order={3}
        project={
          Object {
            "avatarUrl": null,
            "bugs": 13,
            "codeSmells": 5,
            "coverage": 0,
            "duplications": 0,
            "gateStatus": "OK",
            "key": "example-js-2",
            "languages": Array [
              "go",
              "js",
              "php",
              "py",
            ],
            "maintainabilityRating": 1,
            "name": "example-js",
            "ncloc": 80,
            "organizationKey": "autoscan",
            "organizationName": "AutoScan",
            "reliabilityRating": 1,
            "securityRating": 1,
            "vulnerabilities": 0,
          }
        }
        viewable={false}
      />
      <ProjectCard
        key="example-js-3"
        order={4}
        project={
          Object {
            "avatarUrl": null,
            "bugs": 13,
            "codeSmells": 5,
            "coverage": 0,
            "duplications": 0,
            "gateStatus": "OK",
            "key": "example-js-3",
            "languages": Array [
              "go",
              "js",
              "php",
              "py",
            ],
            "maintainabilityRating": 1,
            "name": "example-js",
            "ncloc": 80,
            "organizationKey": "autoscan",
            "organizationName": "AutoScan",
            "reliabilityRating": 1,
            "securityRating": 1,
            "vulnerabilities": 0,
          }
        }
        viewable={false}
      />
    </div>
  </div>
  <button
    className="js-next sc-project-button"
    onClick={[Function]}
    type="button"
  >
    <img
      alt=""
      src="/images/sonarcloud/chevron-right.svg"
    />
  </button>
</div>
`;
